types/src/bytesrepr.rs

```diff
@@ -772,6 +836,213 @@ impl<T: FromBytes, E: FromBytes> FromBytes for Result<T, E> {
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        self.octets().write_bytes(writer)
+    }
+}
+
+#[cfg(feature = "std")]
+impl FromBytes for Ipv4Addr {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (octets, remainder): ([u8; IPV4_SERIALIZED_LENGTH], &[u8]) =
+            FromBytes::from_bytes(bytes)?;
+        Ok((octets.into(), remainder))
+    }
+}
+
+#[cfg(feature = "std")]
+impl ToBytes for Ipv6Addr {
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        let mut buf = allocate_buffer(self)?;
+        self.write_bytes(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn serialized_length(&self) -> usize {
+        IPV6_SERIALIZED_LENGTH
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        self.octets().write_bytes(writer)
+    }
+}
+
+#[cfg(feature = "std")]
+impl FromBytes for Ipv6Addr {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (octets, remainder): ([u8; IPV6_SERIALIZED_LENGTH], &[u8]) =
+            FromBytes::from_bytes(bytes)?;
+        Ok((octets.into(), remainder))
+    }
+}
+
+#[cfg(feature = "std")]
+impl ToBytes for IpAddr {
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        let mut buf = allocate_buffer(self)?;
+        self.write_bytes(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn serialized_length(&self) -> usize {
+        match self {
+            IpAddr::V4(_) => U8_SERIALIZED_LENGTH + IPV4_SERIALIZED_LENGTH,
+            IpAddr::V6(_) => U8_SERIALIZED_LENGTH + IPV6_SERIALIZED_LENGTH,
+        }
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        match self {
+            IpAddr::V4(addr) => {
+                writer.push(IP_V4_TAG);
+                addr.write_bytes(writer)
+            }
+            IpAddr::V6(addr) => {
+                writer.push(IP_V6_TAG);
+                addr.write_bytes(writer)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl FromBytes for IpAddr {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (tag, remainder) = u8::from_bytes(bytes)?;
+        match tag {
+            IP_V4_TAG => {
+                let (addr, rem) = Ipv4Addr::from_bytes(remainder)?;
+                Ok((Self::V4(addr), rem))
+            }
+            IP_V6_TAG => {
+                let (addr, rem) = Ipv6Addr::from_bytes(remainder)?;
+                Ok((Self::V6(addr), rem))
+            }
+            _ => Err(Error::Formatting),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl ToBytes for SocketAddrV4 {
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        let mut buf = allocate_buffer(self)?;
+        self.write_bytes(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn serialized_length(&self) -> usize {
+        U16_SERIALIZED_LENGTH + IPV4_SERIALIZED_LENGTH
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        self.port().write_bytes(writer)?;
```

Comment:
Is there a reason we are writing these in what, at least to me, is "reverse" order? The implementation is platform specific of course, but I assume most people expect these to be in "ip, port" order?

Reply:
No reason other than my preference of serializing primitives first. This only makes a difference in the intermediary bytes representation, which I'm still not sure it's [relevant](https://github.com/casper-network/casper-node/pull/2922#discussion_r864859094).

Reply:
I usually view these things with rose colored glasses of us eventually getting to use a derive macro for these things. If that were the case, I would stick to the field order declared in struct itself, which defines at least one possible canonical ordering.

Reply:
I refactored it.